Add caching for `_rename_without_collisions`

hsjts0u opened the pull request:
Fixes #158357

hsjts0u committed:
Use cache

hsjts0u committed:
Lint and add patch

pianpwk commented on torch/_export/utils.py:
```diff
@@ -874,6 +900,26 @@ def _name_hoo_subgraph_placeholders(gm: torch.fx.GraphModule) -> None:
+                len(find_available[prefix]) >= 2
+                and find_available[prefix][0] + 1 == find_available[prefix][1]
+            ):
+                find_available[prefix].pop(0)
```
Hm, what's this routine for?

hsjts0u commented on torch/_export/utils.py:
```diff
@@ -255,30 +257,54 @@ def _get_shape_env_from_gm(gm: torch.fx.GraphModule):
+            find_available[prefix][0] + 1 == find_available[prefix][1]
+            or find_available[prefix][0] == find_available[prefix][1]
+        ):
+            find_available[prefix].pop(0)
```
I did try naively incrementing tracking the max index so far but that did not pass some of the tests so I opted for this version that mimics the results of the previous algorithm more closely. Should we go we the naive implementation instead? 

Dumb question here, but what tests should I run and is there a fast way to run a whole collection of tests locally instead of having to call `python {path to test file}` for every test I think is important? And if there are failures that are simply expected string mismatches what can I do to quickly address these issues?

pianpwk replied:
No, not dumb at all. Anything under `test/export` is most relevant, and if the naive way just leads to minor naming changes, we can update the expected results. For a more comprehensive test, someone needs to approve a CI run on each commit (just did, you should see some failures). I think there's a way to approve it for all commits going forward though.

hsjts0u committed:
Use a naive implementation

Rename the nodes so that anything that is renamed is strictly
increasing. This means gaps in suffixes are allowed.

hsjts0u committed:
Address linter issues

hsjts0u commented:
@pianpwk Can you take another look when you have time? Thanks
